Add a hyper-parameter tuning module

micyril opened the pull request.

micyril committed:
Add a cross-validation wrapper for optimizers

micyril committed:
Add grid-search optimization

micyril committed:
Add hyper-parameter tuning

micyril committed:
Add support for metrics that should be maximized

micyril committed:
Add interface for accessing the best model

micyril committed:
Fix lambda sets initialization

rcurtin commented on src/mlpack/core/hpt/bind.hpp:
```diff
@@ -0,0 +1,111 @@
+ * Mark the given argument as one that should be bound. It can be applied to
+ * arguments that are passed to the Optimize method of HyperParameterTuner.
+ *
+ * The implementation avoids data coping. If the passed argument is an l-value
```
Minor misspelling---`data copying` not `data coping`.

rcurtin commented on src/mlpack/core/hpt/hpt.hpp:
```diff
@@ -0,0 +1,268 @@
+ * // Using 80% of data for training and remaining 20% for assessing MSE.
+ * double validationSize = 0.2;
+ * HyperParamterTuner<LinearRegression, MSE, SimpleCV> hpt(validationSize, data,
+ *     responses);
```
Tiny misspelling here---should be `HyperParameterTuner`.

rcurtin commented on src/mlpack/core/hpt/hpt.hpp:
```diff
@@ -0,0 +1,268 @@
+ * // and 1.0.
+ * arma::vec lambdas{0.0, 0.001, 0.01, 0.1, 1.0};
+ * double bestLambda;
+ * std::tie(bestLambda) = hpt.Optimize(lambdas);
```
Should we point out here that all of the parameters for the `MLAlgorithm` should be specified?  (In the case of `LogisticRegression` there is only one parameter, lambda.)

micyril replied:
It's not completely true. E.g. for `SoftmaxRegression` we have the following constructor:

```
  template<typename OptimizerType = mlpack::optimization::L_BFGS>
  SoftmaxRegression(const arma::mat& data,
                    const arma::Row<size_t>& labels,
                    const size_t numClasses,
                    const double lambda = 0.0001,
                    const bool fitIntercept = false,
                    OptimizerType optimizer = OptimizerType());
```
During hyper-parameter optimization we can only specify values for `lambda` and skip specifying values for `fitIntercept`. Then for `fitIntercept` the standard value (false) will be used.

rcurtin replied:
Ah, right, so then is this correct: for a given `MLAlgorithm` we must specify all of the required parameters (but parameters with defaults don't need to be specified).

micyril replied:
Yes, it is true. I will mention it in the documentation.

rcurtin commented on src/mlpack/core/hpt/hpt.hpp:
```diff
@@ -0,0 +1,268 @@
+ * To construct a HyperParameterTuner object you need to pass the same arguments
+ * as for construction of an object of the given CV class. For example, we can
+ * use the following code to try to find a good lambda value for
+ * LinearRegression.
```
Do you mean `LogisticRegression`?  `LinearRegression` has no penalty parameter (as currently implemented).

micyril replied:
I had the following `LinearRegression` constructor in my mind:
```
  LinearRegression(const arma::mat& predictors,
                   const arma::rowvec& responses,
                   const double lambda = 0,
                   const bool intercept = true);
```
Isn't lambda a penalty here?

rcurtin replied:
You're right, sorry about that.  Maybe I was looking at some branch or old version that did not have `lambda`.

rcurtin commented on src/mlpack/core/hpt/hpt.hpp:
```diff
@@ -0,0 +1,268 @@
+ * @tparam WeightsType The type of weights (should be passed when weighted
+ *     learning is supported, and the weights type is a template parameter in
+ *     Train methods of the given MLAlgorithm; arma::vec will be used
+ *     otherwise).
```
Just to make sure I am understanding right, I think you said you are planning on writing a longer tutorial for CV and HPT and how to use them?

micyril replied:
When we originally discussed the idea of writing tutorials for CV and HPT, it popped up when SimpleCV referred to CVBase in its constructor. Now the situation is different. Is there a need to write tutorials for the CV and HPT modules? If so, we can do that.

rcurtin replied:
Up to you---adding comprehensive tutorials is always nice because it can allow the user some nice examples and demonstrations of how it works that they can copy-paste and adapt into their own work.  But it is more important to get the functionality working.  So I would say, if you want to, maybe we can do this later if there is time during GSoC.

micyril replied:
Ok

rcurtin commented on src/mlpack/core/hpt/hpt.hpp:
```diff
@@ -0,0 +1,268 @@
+  MLAlgorithm bestModel;
+
+  /**
+   * The set of methods to initialize the GridSearh optimizer. We basically need
```
Misspelling: `GridSearch`.

rcurtin commented on src/mlpack/tests/hpt_test.cpp:
```diff
@@ -0,0 +1,219 @@
+  bool useCholesky = false;
+  arma::vec lambda1Set =
+      arma::join_cols(arma::vec{0}, arma::logspace<arma::vec>(-3, 2, 6));
+  std::array<double, 4> lambda2Set{{0.0, 0.05, 0.5, 5.0}};
```
Just to check, it is probably best if this test has optimal lambda values not at either end of the grid search (so, expectedLambda1 != 0, expectedLambda1 != 100, expectedLambda2 != 0.0, and expectedLambda2 != 5.0).  Basically what I mean is that the optimization may be very "easy" if the solution is at the boundary, so it would be better if the solution is somewhere in the "middle" of the space.  Do you know what the expectedLambda1 and expectedLambda2 values are?

micyril replied:
When I run tests, the values of `expectedLabmda1` and `expectedLambda2` are equal to 0.1 and 0.0 respectively in `GridSearchTest`, and 0.01 and 0.05 respectively in `HPTTest`. I guess we can set some seed for the random generator to always get values somewhere in the "middle" of the space.

rcurtin replied:
Ah, I think those values are okay.  As long as we are not fully at the edge of the grid I think it's fine.  My concern would just be that if the results are `0` for both lambdas in either test, then it is possible that the grid search failed to perform completely and simply returned the first result.  But based on what you say, that is not the case, so I think this is fine.  If you want to try and modify so that `lambda2` is not `0.0` for `GridSearchTest`, feel free, but don't feel obligated.

micyril replied:
if we do `arma::arma_rng::set_seed(11);`, then `expectedLabmda1` and `expectedLambda2` are equal to 0.01 and 0.05 respectively.

rcurtin replied:
Ok, that sounds good.  We should avoid setting the random seed in the tests though (and for setting the seeds, it is generally better to use `math::RandomSeed()` instead, so that the mlpack RNGs are using just one seed too).

micyril replied:
So, should I revert the commit ["Make optimal lambdas in the "middle" of the space?"](https://github.com/micyril/mlpack/pull/2/commits/c6f897f05dd22cbe2b649b7a11a37c7f5b7d9693)?

rcurtin replied:
Ah, it is not a great idea to rely on random seeds for a particular test result in general.  On a different system the results you have may not apply.  Instead it might be better to use a different seed each run:

```
math::RandomSeed(std::time(NULL));
```

and then have the test print the found optimal lambdas for several runs:

```
for i in `seq 1 20`; do bin/mlpack_test -t HPTTest/GridSearchTest; sleep 1; done
```

If the expected lambdas are typically (but not always) nonzero values like you said, I think that is just fine.  But in any case, we shouldn't set the seed in the tests, and that means that we also can't test for specific values of lambda, so I think that yes, the commit should be reverted.

micyril replied:
I have made the experiment - in 17 out of 20 runs at least one lambda value was non-zero.

rcurtin commented on src/mlpack/tests/hpt_test.cpp:
```diff
@@ -0,0 +1,219 @@
+
+  // Making sure that the assumption above is true.
+  SimpleCV<LogisticRegression<>, Accuracy>
+    cv(validationSize, doubledXs, doubledYs);
```
Minor style issue---this line should be indented four spaces, not two.

rcurtin commented on src/mlpack/core/hpt/hpt.hpp:
```diff
@@ -0,0 +1,268 @@
+ * lambda1 and lambda2 values for LARS.
+ *
+ * @code
+ * HyperParamterTuner<LARS, MSE, SimpleCV> hpt2(validationSize, data,
```
Minor misspelling here too.

rcurtin commented on src/mlpack/core/hpt/hpt.hpp:
```diff
@@ -0,0 +1,268 @@
+               typename std::tuple_element<I, ArgsTuple>::type>::value>,
+           typename = void>
+  inline void InitGridSearch(const ArgsTuple& args,
+                             Collections... collections);
```
Like in another PR, maybe it would be cleaner to use SFINAE as a function argument here instead of as a template argument.  But it is also possible that there's a reason this doesn't work that I have overlooked...

micyril replied:
If I remember it right, we can't do something like this
```
  inline void InitGridSearch(const ArgsTuple& args,
                             Collections... collections,
                             std::enable_if_t<...>* = 0);
```
since compilers will not match passed collections with the `collections` parameter, rather they will try to match them against the last argument.

micyril replied:
Actually, we can use SFINAE as a function argument with variadic templates too by explicitly specifying variadic template arguments during calls. Let me know if you want me to do it (I can also change [SimpleCV/CVBase](https://github.com/mlpack/mlpack/pull/1044) code, since I use SFINAE there in the same manner).

micyril replied:
I reviewed the `SimpleCV`/`CVBase` code - I think it doesn't make sense to use SFINAE as a function argument there since anyway we are obligated to use a template bool variable like `Enabled` to condition on (at least I don't know any other solution).

rcurtin replied:
Ah, that is what I missed---variadic templates do not allow the argument approach to work.  In that case I think what you are doing is fine, but you could also do return-type SFINAE perhaps, like:

```
template<....>
std::enable_if_t<condition, void>
Function(...);
```

I think it's ok to do whatever you think is best, I just wanted to point this out as a possibility in case you think it gives cleaner code.

micyril replied:
Return-type SFINAE doesn't help much since we can't put everything here into one condition (if we do compilers dosn't compile it). So, if we don't want explicitly specify variadic templates, I guess it's better to stay it as it is.

rcurtin replied:
`std::enable_if_t<condition1 && condition2 && condition3>` fails to compile?  I'm surprised to hear that.  In either case, I agree, it's fine as is.

micyril replied:
In our case `condition2` contains the statement `std::tuple_element<I, ArgsTuple>::type>::value`, which throws a static error if the value I turns out to be bigger than the tuple size minus 1. My compiler (clang) doesn't skip this error.

rcurtin commented on src/mlpack/core/hpt/hpt_impl.hpp:
```diff
@@ -0,0 +1,246 @@
+                         WeightsType>::InitCVFunctionAndOptimize(
+    const ArgsTuple& /* args */,
+    arma::mat& bestParams,
+    BoundArgs... boundArgs)
```
Hmm, so maybe this is a comment without a good solution.  I keep reading the term `BoundArgs` and thinking that it refers to arguments that are bounded (like a `math::Range` or something), but it is actually referring to arguments that were passed with `Bind`.  `BoundArgs` would be the correct word to use for an argument that has been bound with `Bind`, but there is still some ambiguity here and at least for me it makes the code a little confusing to read (especially if later the code is modified to support bounded parameters).  Do you think maybe we can come up with a better name for `Bind` that would solve this problem, or a better name for `BoundArgs`?  Maybe `Fixed` would be a better word to use, I am not sure.

micyril replied:
It works in the following way. If a user wants to bind some hyper-parameters to some particular values, he/she uses the `Bind` function for that. This `Bind` function returns objects of the type `PreBoundArg`, which are basically wrapped input arguments. We can think about it like the `Bind` function "marks" arguments that should be bound. In order to be completely "bound", information about argument position should be presented. This information is added by using methods with the name `InitCVFunctionAndOptimize`. Eventually "bound args" are passed to `CVFunction`, for which information about the values and their positions in the argument list is sufficient to bind the hyper-parameters to the values.

When you suggested `Fixed`, did you mean using `FixedArgs` instead of `BoundArgs` and preserving other names as they are now?

rcurtin replied:
Right, I understand how `Bind` works.  My suggestion is just that `BoundArgs` can be a confusing term when reading the code.  So possibly instead of `Bind` and `BoundArgs`, `Fixed` and `FixedArgs` might be names that are less confusing.  If you like the name `BoundArgs` then I think it's ok to stick with it, but be sure to add some comments pointing out that `BoundArgs` refer to arguments that have been bound to one specific fixed value with `Bind()`, otherwise in the future some maintainer might get confused. :)

micyril replied:
I'm totally ok with `Fixed` and `FixedArgs`, so I will change the names.

micyril replied:
I would like to make sure - should we prefer the name `Fixed` rather than `Fix` for the function that was called `Bind`?

rcurtin replied:
Yeah, I think `Fixed()` is a better name to use than `Fix()` here.

rcurtin commented on src/mlpack/core/hpt/hpt_impl.hpp:
```diff
@@ -0,0 +1,246 @@
+         typename MatType,
+         typename PredictionsType,
+         typename WeightsType>
+template<size_t I, class ArgsTuple, class... BoundArgs, class, class>
```
Another thought about SFINAE in function arguments vs. template arguments is that here, to understand which overload I am looking at, I have to go back to the header file, count the arguments, and cross-reference with this, which is a little bit inconvenient.

rcurtin commented on src/mlpack/core/hpt/cv_function.hpp:
```diff
@@ -0,0 +1,143 @@
+
+  //! The used machine learning algorithm.
+  using MLAlgorithm = typename
+      std::remove_reference<decltype(std::declval<CVType>().Model())>::type;
```
If `CVType` exposes its own internal `MLAlgorithm` type, this could be a lot simpler.  That's just an idea, I'm indifferent to what you choose to do here.

micyril replied:
We can also make `HyperParameterTuner` pass the `MLAlgorithm` type to `CVFunction`. It will be simpler too, and without making additional requirements for `CVType`. What do you think?

rcurtin replied:
I think that's fine.  `CVFunction` is very specific to the hyper-parameter tuner and cross-validation, so I don't see a regular user needing to use that class.  If you think passing `MLAlgorithm` to `CVFunction` is better, that sounds good to me. :)

rcurtin commented on src/mlpack/core/hpt/cv_function.hpp:
```diff
@@ -0,0 +1,143 @@
+   * rather than an element of parameters at the position PIndex.
+   */
+  template<size_t BAIndex,
+           size_t PIndex,
```
I know it is a little more verbose, but can you use more explicit names than `BAIndex` and `PIndex`?  I think these are `BoundArgIndex` and `ParamIndex` but I am not sure.  I think the code would be easier to understand if I was sure of what the arguments here were.  Adding some documentation in the comment could be another way to solve this.

rcurtin commented on src/mlpack/core/hpt/cv_function_impl.hpp:
```diff
@@ -0,0 +1,118 @@
+  // Change the best model if we have got a better score, or if we probably
+  // have not assigned any valid (trained) model yet.
+  if (bestObjective > objective ||
+      bestObjective == std::numeric_limits<double>::max())
```
I think you can simplify this, `if (objective < bestObjective)` would be sufficient to also capture the case where `bestObjective == DBL_MAX`.

micyril replied:
Do you mean to use `DBL_MAX` instead of `std::numeric_limits<double>::max()`?

rcurtin replied:
No, I was being lazy---it is quicker to type `DBL_MAX` than `std::numeric_limits<double>::max()` and I was using a phone so I did not want to type too much. :)

The idea of what I was saying though, is that there is no need to check if `bestObjective == std::numeric_limits<double>::max()`; if that condition holds, then it will always be true that `objective < bestObjective` (...assuming that `objective` is not `std::numeric_limits<double>::max()`, but even if that is true, that corner case can be handled by relaxing the conditional to `if (objective <= bestObjective)` and thereby taking the better objective even when it is the same.).  Let me know if I can clarify further.

micyril replied:
There are some trade-offs: we can use less lines (and CPU cycles) for condition checking, but we potentially will do more often the `if` body. Also the implementation with one condition is inconsistent with `GridSearch` - as a result we can provide a model trained with some different hyper-parameters than ones returned from `Optimize`.

rcurtin replied:
Ah, I did not realize relaxing the condition made it inconsistent.  In that case it seems like we need to leave it as-is.

rcurtin commented on src/mlpack/core/hpt/cv_function_impl.hpp:
```diff
@@ -0,0 +1,118 @@
+    const Args&... args)
+{
+  return Evaluate<BAIndex, PIndex + 1>(
+      parameters, args..., parameters(PIndex, 0));
```
Another option here is to access `parameters` one-dimensionally, with `parameters[PIndex]` not `parameters(PIndex, 0)`.  I am not sure which you think is more clear; I'm indifferent.  I just wanted to point out the possibility. :)

micyril replied:
I assumed that arma::mat require to pass two indexes. For this concrete case I'm also indifferent, so I guess we can stay it as it is.

rcurtin reviewed (commented):
Sorry that this review took me so long to do.  I know there are a lot of comments but I think mostly they are pretty minor; I hope they are helpful.  To me it is clear you have worked really hard on this and I think it looks great---I am really excited about getting this merged into the codebase.

One larger discussion I would like to have is about `GridSearch`.  For some other work I have been doing, I need to optimize functions that have both discrete and continuous variables.  In that case I realized the best solution was to have the optimizer take a `data::DatasetInfo`.  I think that here we could do a similar thing: the `GridSearch` optimizer is a discrete optimizer.  So suppose that we adapted the API to

```
GridSearch(const data::DatasetInfo& info);
```

and then the `info` object would hold the number of categorical variables to try (`info.NumMappings(i)`).  Then, the `CVFunction` would map from the given parameters to the correct values.  Before I write too much about that, I wanted to ask, what do you think?  I think a refactoring of this type could make it a lot easier to eventually expand the hyper-parameter tuner to other optimizers than just grid search, and also allow continuous search for parameters.

micyril commented:
Hi, Ryan. I don't think that I completely understand why this refactoring makes sense, but there are some of my thoughts regarding the discussion. `CVFunction` was designed to adapt the interface of cv objects to the one that can be utilized by mlpack optimizers. From this prospective `CVFunction` is not aware by which optimizer it's going to be used. If I understand you right, you suggest to make the assumption that `CVFunction` is going to be used only by `GridSearch` (since only it is going to use `const data::DatasetInfo& info` and send indexes that should be mapped to appropriate values). How then are we going to add support for `GradientDesent`?

micyril committed:
Fix misspellings

micyril committed:
Fix CVFunctionTest

micyril committed:
Fix style

rcurtin commented:
Ah, actually I was talking about the interaction between `HyperParameterTuner` and `GridSearch`; I wasn't trying to suggest that we make `CVFunction` only work with `GridSearch`.  I think `CVFunction` is just fine---it implements the standard optimizer `FunctionType` interface so there are no problems there.

The `HyperParameterTuner` code is currently dependent very heavily on the `GridSearch` optimizer though: before optimization, `InitGridSearch()` is called to set up the call to the `GridSearch` constructor, which creates a `Collections...` varargs and then calls the constructor:

```
optimizer = Optimizer(collections...);
```

If we want to make the hyper-parameter tuner generic so that we can eventually use gradient descent and whatever else, we have to figure out a unified way to construct optimizers.  For hyper-parameter tuning I think we'll end up with three kinds of parameters to be optimized:

 - **discrete sets**: this is what `GridSearch` currently uses
 - **unbounded continuous variables**: this is what the other mlpack optimizers use
 - **bounded continuous variables**: i.e. maybe we want to say that the `lambda` value can take any value in a `math::Range(0.0, 100.0)`.

For this PR, I think it is fine to only consider discrete sets.  But we should design our abstraction for discrete sets such that it is easier to add the other types later.

In other places in the mlpack codebase (the Hoeffding tree, decision trees, a few other places I think) discrete parameters are accepted.  The difference between discrete and continuous parameters is denoted by the `data::DatasetInfo` object.  So, my suggestion is that instead of using a `Collections...` type to represent the discrete variables to be optimized, we instead use a `data::DatasetInfo` type.

The parameters to be optimized that are discrete will have `DatasetInfo::Type(d)` value equal to `Datatype::categorical`, and the number of possibilities will be found by `DatasetInfo::NumMappings(d)`.  For the `GridSearch` optimizer, then, we could have the following constructor:

```
GridSearch(const data::DatasetInfo& info);
```

and an exception would be thrown if any `info.Type(i)` was equal to `Datatype::numeric` (that is, `GridSearch` would only accept categorical types).  The other functionality of `GridSearch` can stay the same, but maybe the internal details would have to change a little.  The major change would be that instead of building a `Collections...` with the `InitGridSearch()` method, we would be building up a `data::DatasetInfo` object with the relevant information about each dimension.

There may be some minor details I have overlooked, but that would be the basic idea.  What do you think?  From there, it would be easier to start adding support for mixed discrete-continuous optimizers (not in this PR of course) and later add support for bounded continuous variables (also not in this PR).

Sorry to write so much; let me know if I can clarify what I mean.

rcurtin commented on src/mlpack/core/optimizers/grid_search/grid_search_impl.hpp:
```diff
@@ -0,0 +1,72 @@
+      currentParameters(i) = value;
+      Optimize(function, bestObjective, bestParameters, currentParameters,
+          i + 1);
+    }
```
Ah, another small style comment; can you use braces `{}` after the `if` statement to enclose the `for` loop here please?

micyril replied:
I thought that according the guide lines we should skip put braces in such case:

> To conserve lines, refrain from using braces if the if, while, or for statement applies to a single statement.

Isn't `for` operator considered to be a statement?

rcurtin replied:
Technically you are right, but Marcus has asked people to add braces in those cases for clarity, so do you think it is okay to add them here?  If you agree I'll modify the style guide very slightly: "refrain from using braces if the ... statement applies to a single-line statement".

micyril replied:
Ok, I will add the change.

micyril commented:
> The HyperParameterTuner code is currently dependent very heavily on the GridSearch optimizer though: before optimization, InitGridSearch() is called to set up the call to the GridSearch constructor, which creates a Collections... varargs and then calls the constructor:

I was planning to generalize it in the following way: if arguments passed to the Optimize method of HPT are of arithmetic types, then they should be used as initial values for the Optimize method of the chosen optimizer (as in the case of `GradientDesent`); otherwise they should be used as constructor arguments for the chosen optimizer (as in the case of `GridSearch`).

> The parameters to be optimized that are discrete will have DatasetInfo::Type(d) value equal to Datatype::categorical, and the number of possibilities will be found by DatasetInfo::NumMappings(d)

But how will `GridSearch` know what exact values should be tried for hyper-parameters if it is supposed to take only a `DatasetInfo` object? Can `DatasetInfo` store this information?

micyril committed:
Extend documentation

micyril committed:
Fix typo

rcurtin commented:
> I was planning to generalize it in the following way: if arguments passed to the Optimize method of HPT are of arithmetic types, then they should be used as initial values for the Optimize method of the chosen optimizer (as in the case of GradientDesent); otherwise they should be used as constructor arguments for the chosen optimizer (as in the case of GridSearch).

Ok, that seems like a reasonable strategy to me.  I have some ideas for how we can support bounded continuous variables, but we can talk about that later---there is no need for that in this PR.

> But how will GridSearch know what exact values should be tried for hyper-parameters if it is supposed to take only a DatasetInfo object? Can DatasetInfo store this information?

The `DatasetInfo` class is essentially a class that holds a mapping between some original type and some new type.  (Aside: right now the original type is always std::string, but it would be an easy change to adapt and templatize.  For holding arbitrary types, boost::any could be used.  For the rest of the discussion though, please overlook this current restriction and assume we (or I) can easily refactor to make the code ideas below work.)

When we have some set of parameters we want to optimize, like

```
arma::vec lambdas { 0.0, 0.1, 0.5, 1.0 };
```

this is four separate discrete values.  Assuming that we are only optimizing the lambda parameter, so that the parameters matrix is one-dimensional, we can use `DatasetInfo` to represent the mapping:

```
DatasetInfo di(1); // Only one dimension.
// Technically this only accepts strings for now.  But like I said in the parenthetical earlier it would be easy
// to refactor.  So, this snippet assumes that it has been refactored, but I left the name of the method
// intact.
for (size_t i = 0; i < lambdas.n_elem; ++i)
  di.MapString(lambdas[i], 0 /* map in dimension 0 */);
```

Now we have an object that holds the mapping completely, and our set of lambdas is represented as a discrete (categorical) dimension.  Thus,

```
di.Type(0) == Datatype::categorical
```

and we can get the number of possible values the dimension can take:

```
di.NumMappings(0) == 4 // this is equal to lambdas.n_elem
```

So, `GridSearch` would receive this `DatasetInfo` object that encoded all of the possible values for hyperparameters during the search.  For our simple example search then, it would need to search with only four parameters matrices:

```
[0]
[1]
[2]
[3]
```

and then `CVFunction` can unmap from the parameter matrix back to the original value with the `DatasetInfo` object (which I guess should be passed to the constructor or something):

```
// Get the original value back.
const double lambdaValue = di.UnmapString(parameters[0], 0 /* dimension 0 */);
```

I think this approach is more consistent with the rest of mlpack, which would be important if we ever decide to add more discrete optimizers for other tasks (for instance, simulated annealing is a perfect candidate for discrete/categorical optimization; our current implementation only has continuous/numeric support).  This is also the approach I am using for some prototype code I have which does mixed categorical/numeric optimization.

Does what I am writing make sense?  Please let me know and I can elaborate further if needed.  There are some additional details about how to use `DatasetInfo` with arbitrary types that I have not written about here so that the response is not too long. :)

micyril commented:
> // Get the original value back.
> const double lambdaValue = di.UnmapString(parameters[0], 0 /* dimension 0 */);

I'm not sure what `parameters[0]` is. Is it the index of the original value in the collection `lambdas`?

My probably last concern about using `DatasetInfo` in the suggest way is the following. Is it really useful to make `HyperParameterTuner` create a `DatasetInfo` object that should be passed to the constructor of the chosen optimizer? Maybe `HyperParameterTuner` should only pass the arguments (e.g. collections and ranges) which can be used for `DatasetInfo` construction if it is really needed? I say it because it is probably not that useful for `GridSearch` and it probably makes the `GridSearch` interface less convenient. But, anyway, you probably has more far going vision than me, and according to it forcing to use a `DatasetInfo` object  as a constructor argument for optimizers is really convenient.

rcurtin commented:
`parameters[0]` is the element passed to `CVFunction::Evaluate()`.  In my example it would indeed be the index of the original value in the collection `lambdas`.

I agree that taken alone, forcing `HyperParameterTuner` to generate a `DatasetInfo` object which it in turn passes to the `GridSearch` optimizer is not that useful by itself.  But when we start to consider other optimizers for `HyperParameterTuner`, which may accept continuous parameters, discrete parameters, or a combination of both, then the refactoring becomes useful.  Using `DatasetInfo` gives us a concrete abstraction through which we can support different optimizers for the hyper-parameter tuner.  Without `DatasetInfo` (or something very similar to it), to me it is not clear how we can support all of these different types of parameters to be optimized in a way that "matches" the rest of mlpack.

If all of that makes sense, let me know, and I'll provide the necessary patch for `DatasetInfo` so that it can handle non-string types.  If not, let me know, I am happy to discuss more. :)

micyril commented:
`Without DatasetInfo (or something very similar to it), to me it is not clear how we can support all of these different types of parameters to be optimized in a way that "matches" the rest of mlpack.`

I thought that we could pass the same argument to optimizer constructors as ones that would be passed to the Optimize method of `HyperParameterTuner` - these are collections in the case of discrete values, and whatever can be designed in the case of bounded continuous variables. I just wanted to make my opinion clearer, so please don't consider it as disapproval. I'm ready to take for granted that it is a way that matches the rest of mlpack.

> I'll provide the necessary patch for DatasetInfo so that it can handle non-string types

Yes, it would be nice.

micyril commented:
There is yet another concern from my side. I just started to think how we can support bounded continuous variables. From user perspective I can imagine the following interface.
```
// Suppose we have already created hpt object of the right type and want to
// optimize LogisticRegression
mlpack::math::Range lambdaRange(0.0, 10.0);
double bestLambda;
std::tie(bestLambda) = hpt.Optimize(lambdaRange);
```
My initial idea was to delegate handling all these parameters to the choosen optimizer (by calling its constructor). Now, with the `DatasetInfo`-based implementation in mind, if we just try to initialize a `DatasetInfo` object with `lambdaRange`, assuming it is a collection, we will fail. Maybe you have some thoughts how it is going to supported, so we can probably make the current `HyperParameterTuner` implementation easier for adding this support in feature.

micyril committed:
Make optimal lambdas in the "middle" of the space

micyril committed:
Add passing MLAlgorithm type to CVFunction

micyril committed:
Add braces around multiline statement

micyril committed:
Use more verbose names for BAIndex and PIndex

rcurtin commented:
So there is a difference between the user interface to the hyper-parameter tuner, and the user interface for discrete/continuous optimizers.  I think the user interface you are proposing for `HyperParameterTuner` is just fine---I am more concerned with the lower-level optimizer interface.  So having `HyperParameterTuner` accept a `math::Range()` for a bounded continuous variables is the right idea in my opinion.

Delegating all of the handling of parameters to the chosen optimizer is somewhat problematic because it's not likely to work with existing optimizers unless we refactor them all to accept a `Collections...`, which would be a large amount of work (and then the interface may be confusing for users).  Instead, it would be less work to refactor them to accept a `DatasetInfo`.  For existing continuous-only optimizers (like `SGD`, `SA`, etc.), we simply need to add a constructor that takes a `DatasetInfo` and ensures that all dimensions are of type `Datatype::numeric`.

The idea I had played with is that we have three classes of optimizers with the following `Optimize()` signatures:

```
// Optimize() functions for continuous-only optimizer (the current mlpack default).
void ContinuousOnlyOptimizer::Optimize(const arma::mat& parameters);

// Optimize() functions for optimizers that take discrete arguments (including GridSearch,
// but GridSearch has the restriction that _all_ arguments must be discrete).
void DiscreteOptimizer::Optimize(const arma::mat& parameters, const data::DatasetInfo& info);

// Optimize() functions for optimizers that take bounded arguments.  I think either of the following
// overloads is fine, depending on whether the optimizer supports discrete arguments.
void BoundedOptimizer::Optimize(const arma::mat& parameters,
                                const data::DatasetInfo& info,
                                const ConstraintsType& constraints);

void BoundedOptimizer::Optimize(const arma::mat& parameters, const ConstraintsType& constraints);
```

Then the `ConstraintsType` class (which may or may not be a template argument, I am not sure if it is needed yet) can implement the following interface:

```
// Get whether or not the argument is constrained.
bool ConstraintsType::IsConstrained(const size_t index);

// Return the range in which an argument is constrained.
const math::Range Range(const size_t index);
```

I have already written a basic `Constraints` class in some other code that I haven't pushed to Github yet.  So, when `HyperParameterTuner` receives a `math::Range` argument, we can first check to see if the chosen optimizer supports constrained optimization, and if it does, then we can assemble a `Constraints` object accordingly and pass that to the optimizer along with the `DatasetInfo`.

If you see a problem with this general abstraction, let me know---I think it will work fine, but perhaps there is some case I overlooked.  If this abstraction works ok for you, then I guess we can adapt `GridSearch` slightly to use it and merge this PR.  Alternately, if it is a lot of work to refactor `GridSearch`, we can accept this as-is for now and then later do the refactoring.

micyril commented:
> Delegating all of the handling of parameters to the chosen optimizer is somewhat problematic because it's not likely to work with existing optimizers unless we refactor them all to accept a Collections..., which would be a large amount of work (and then the interface may be confusing for users).

Actually, it wasn't supposed to refactor the existing optimizers. Rather it was supposed to apply the strategy that I described earlier:

> if arguments passed to the Optimize method of HPT are of arithmetic types, then they should be used as initial values for the Optimize method of the chosen optimizer

Anyway, now I like your design too, since it makes it easier to implement simultaneous optimization of continuous and discrete parameters. 

> // Optimize() functions for optimizers that take discrete arguments (including GridSearch,
> // but GridSearch has the restriction that _all_ arguments must be discrete).
> void DiscreteOptimizer::Optimize(const arma::mat& parameters, const data::DatasetInfo& info);

So, maybe we implement this new standard interface in `GridSearch` right now, skipping passing `DatasetInfo` into the constructor? I guess we should also take a function to optimize:
```
template<typename FunctionType>
void DiscreteOptimizer::Optimize(FunctionType& function, const arma::mat& parameters, const data::DatasetInfo& info);
```

rcurtin commented:
Ah yes, I forgot the `FunctionType` in those `Optimize()` signatures.  I agree, refactoring `GridSearch` to have the `Optimize()` function you just described would be good.

> Actually, it wasn't supposed to refactor the existing optimizers. Rather it was supposed to apply the strategy that I described earlier:
>
>> if arguments passed to the Optimize method of HPT are of arithmetic types, then they should be used as initial values for the Optimize method of the chosen optimizer

Ah, I see what you mean now.  I think with this new strategy we can still use the arithmetic types as the initial points.  When a user passes a `math::Range()` to the `HyperParameterTuner`, I guess maybe we can use the midpoint of the range as an initial point.

In either case, it sounds like we agree on a good abstraction for the optimizers, so we can work out the other details as we go.  If you think of any other problems (or any other solutions that you like more), let me know.

micyril commented:
> In either case, it sounds like we agree on a good abstraction for the optimizers, so we can work out the other details as we go. If you think of any other problems (or any other solutions that you like more), let me know.

No, now it seems to be the best solution so far. So, waiting implementation of the missing functionality of `DatasetInfo` by you.

micyril committed:
Use the word "fixed" for "bound"

rcurtin commented:
I think this patch will work for `DatasetMapper`:
https://gist.github.com/rcurtin/0628ebb53f349ce60a6dc6099902005d

You can take a look at the new tests in `imputation_test.cpp` to see how it can be used.  Maybe the API could be improved a little, but at least it works for now (for instance, maybe `MapString` should be renamed now).

micyril committed:
Revert "Make optimal lambdas in the "middle" of the space"

This reverts commit c6f897f05dd22cbe2b649b7a11a37c7f5b7d9693.

micyril commented:
Ryan, it looks like you forgot to include into the patch the file `mlpack/core/data/map_policies/map_all.hpp`

rcurtin commented:
Ah, sorry about that.  `map_all.hpp` is not actually necessary---I forgot to remove the `#include` from `imputation_test.cpp`.  If you remove that line, I think it should work fine (let me know if not).

micyril commented:
If I apply the patch to the master branch, I get the following error:
```
error: patch failed: src/mlpack/core/data/dataset_mapper.hpp:51
error: src/mlpack/core/data/dataset_mapper.hpp: patch does not apply
```
If apply the patch to the hpt branch, during compilation I get the following error:
```
In file included from /Users/kirill/Evial/Current/mlpack/src/mlpack/tests/imputation_test.cpp:14:
In file included from /Users/kirill/Evial/Current/mlpack/src/mlpack/core.hpp:234:
In file included from /Users/kirill/Evial/Current/mlpack/src/mlpack/core/util/cli.hpp:23:
In file included from /usr/local/include/boost/program_options.hpp:15:
In file included from /usr/local/include/boost/program_options/options_description.hpp:13:
In file included from /usr/local/include/boost/program_options/value_semantic.hpp:14:
In file included from /usr/local/include/boost/lexical_cast.hpp:32:
In file included from /usr/local/include/boost/lexical_cast/try_lexical_convert.hpp:42:
In file included from /usr/local/include/boost/lexical_cast/detail/converter_lexical.hpp:38:
In file included from /usr/local/include/boost/type_traits/has_left_shift.hpp:43:
/usr/local/include/boost/type_traits/detail/has_binary_operator.hpp:158:70: error: use of overloaded
      operator '<<' is ambiguous (with operand types 'std::__1::basic_ostream<char>' and
      'mlpack::data::Datatype')
  ...value = (sizeof(s_check(((make<Lhs>() BOOST_TT_TRAIT_OP make<Rhs>()),make<has_operator>())))==siz...
                               ~~~~~~~~~~~ ^                 ~~~~~~~~~~~
/usr/local/include/boost/type_traits/has_left_shift.hpp:13:27: note: expanded from macro
      'BOOST_TT_TRAIT_OP'
#define BOOST_TT_TRAIT_OP <<
                          ^
/usr/local/include/boost/config/suffix.hpp:394:72: note: expanded from macro 'BOOST_STATIC_CONSTANT'
#     define BOOST_STATIC_CONSTANT(type, assignment) static const type assignment
                                                                       ^~~~~~~~~~
/usr/local/include/boost/type_traits/detail/has_binary_operator.hpp:180:16: note: in instantiation of
      template class 'boost::detail::has_left_shift_impl::operator_exists<std::__1::basic_ostream<char>,
      mlpack::data::Datatype>' requested here
      value = (operator_exists < Lhs, Rhs >::value && operator_returns_Ret < Lhs, Rhs, Ret, opera...
               ^
/usr/local/include/boost/type_traits/detail/has_binary_operator.hpp:208:41: note: in instantiation of
      template class 'boost::detail::has_left_shift_impl::trait_impl1<std::__1::basic_ostream<char>,
      mlpack::data::Datatype, boost::detail::has_left_shift_impl::dont_care, false>' requested here
   BOOST_STATIC_CONSTANT(bool, value = (trait_impl1 < Lhs_noref, Rhs_noref, Ret, BOOST_TT_FORBIDD...
                                        ^
/usr/local/include/boost/type_traits/detail/has_binary_operator.hpp:216:119: note: in instantiation of
      template class 'boost::detail::has_left_shift_impl::trait_impl<std::__1::basic_ostream<char>,
      mlpack::data::Datatype, boost::detail::has_left_shift_impl::dont_care>' requested here
  ...: public integral_constant<bool, (::boost::detail::BOOST_JOIN(BOOST_TT_TRAIT_NAME, _impl)::trait_...
                                                                                                ^
/usr/local/include/boost/test/tools/detail/print_helper.hpp:52:46: note: in instantiation of template
      class 'boost::has_left_shift<std::__1::basic_ostream<char>, mlpack::data::Datatype,
      boost::detail::has_left_shift_impl::dont_care>' requested here
            BOOST_STATIC_ASSERT_MSG( (boost::has_left_shift<std::ostream,T>::value),
                                             ^
/usr/local/include/boost/test/tools/detail/print_helper.hpp:61:24: note: in instantiation of function
      template specialization
      'boost::test_tools::tt_detail::impl::boost_test_print_type<mlpack::data::Datatype>' requested here
                return boost_test_print_type(ostr, r);
                       ^
/usr/local/include/boost/test/tools/detail/print_helper.hpp:90:9: note: in instantiation of function
      template specialization
      'boost::test_tools::tt_detail::impl::boost_test_print_type_impl::operator()<mlpack::data::Datatype>'
      requested here
        boost_test_print_type(ostr, t);
        ^
/usr/local/include/boost/test/tools/detail/print_helper.hpp:216:5: note: in instantiation of member
      function 'boost::test_tools::tt_detail::print_log_value<mlpack::data::Datatype>::operator()'
      requested here
    print_log_value<T>()( ostr, ph.m_t );
    ^
/usr/local/include/boost/test/utils/lazy_ostream.hpp:66:29: note: in instantiation of function template
      specialization 'boost::test_tools::tt_detail::operator<<<mlpack::data::Datatype>' requested here
        return m_prev(ostr) << m_value;
                            ^
/usr/local/include/boost/test/tools/old/impl.hpp:92:50: note: in instantiation of member function
      'boost::unit_test::lazy_ostream_impl<boost::unit_test::lazy_ostream,
      boost::test_tools::tt_detail::print_helper_t<mlpack::data::Datatype>, const
      boost::test_tools::tt_detail::print_helper_t<mlpack::data::Datatype> &>::operator()' requested here
BOOST_PP_REPEAT( BOOST_TEST_MAX_PREDICATE_ARITY, IMPL_FRWD, _ )
                                                 ^
/usr/local/include/boost/preprocessor/repetition/repeat.hpp:29:26: note: expanded from macro
      'BOOST_PP_REPEAT'
# define BOOST_PP_REPEAT BOOST_PP_CAT(BOOST_PP_REPEAT_, BOOST_PP_AUTO_REC(BOOST_PP_REPEAT_P, 4))
                         ^
/usr/local/include/boost/preprocessor/cat.hpp:22:32: note: expanded from macro 'BOOST_PP_CAT'
#    define BOOST_PP_CAT(a, b) BOOST_PP_CAT_I(a, b)
                               ^
/usr/local/include/boost/preprocessor/cat.hpp:29:34: note: expanded from macro 'BOOST_PP_CAT_I'
#    define BOOST_PP_CAT_I(a, b) a ## b
                                 ^
<scratch space>:3:1: note: expanded from here
BOOST_PP_REPEAT_1
^
/Users/kirill/Evial/Current/mlpack/src/mlpack/tests/imputation_test.cpp:282:3: note: in instantiation of
      function template specialization
      'boost::test_tools::tt_detail::check_frwd<boost::test_tools::tt_detail::equal_impl_frwd,
      mlpack::data::Datatype, mlpack::data::Datatype>' requested here
  BOOST_REQUIRE_EQUAL(dm.Type(0), data::Datatype::categorical);
  ^
/usr/local/include/boost/test/tools/old/interface.hpp:154:45: note: expanded from macro
      'BOOST_REQUIRE_EQUAL'
#define BOOST_REQUIRE_EQUAL( L, R )         BOOST_TEST_TOOL_IMPL( 0, \
                                            ^
/usr/local/include/boost/test/tools/old/interface.hpp:65:47: note: expanded from macro
      'BOOST_TEST_TOOL_IMPL'
    BOOST_PP_IF( frwd_type, report_assertion, check_frwd ) (                    \
                                              ^
/usr/local/Cellar/llvm/4.0.1/bin/../include/c++/v1/ostream:208:20: note: candidate function
    basic_ostream& operator<<(bool __n);
                   ^
/usr/local/Cellar/llvm/4.0.1/bin/../include/c++/v1/ostream:211:20: note: candidate function
    basic_ostream& operator<<(int __n);
                   ^
/usr/local/Cellar/llvm/4.0.1/bin/../include/c++/v1/ostream:209:20: note: candidate function
    basic_ostream& operator<<(short __n);
                   ^
/usr/local/Cellar/llvm/4.0.1/bin/../include/c++/v1/ostream:210:20: note: candidate function
    basic_ostream& operator<<(unsigned short __n);
                   ^
/usr/local/Cellar/llvm/4.0.1/bin/../include/c++/v1/ostream:212:20: note: candidate function
    basic_ostream& operator<<(unsigned int __n);
                   ^
/usr/local/Cellar/llvm/4.0.1/bin/../include/c++/v1/ostream:213:20: note: candidate function
    basic_ostream& operator<<(long __n);
                   ^
/usr/local/Cellar/llvm/4.0.1/bin/../include/c++/v1/ostream:214:20: note: candidate function
    basic_ostream& operator<<(unsigned long __n);
                   ^
/usr/local/Cellar/llvm/4.0.1/bin/../include/c++/v1/ostream:215:20: note: candidate function
    basic_ostream& operator<<(long long __n);
                   ^
/usr/local/Cellar/llvm/4.0.1/bin/../include/c++/v1/ostream:216:20: note: candidate function
    basic_ostream& operator<<(unsigned long long __n);
                   ^
/usr/local/Cellar/llvm/4.0.1/bin/../include/c++/v1/ostream:217:20: note: candidate function
    basic_ostream& operator<<(float __f);
                   ^
/usr/local/Cellar/llvm/4.0.1/bin/../include/c++/v1/ostream:218:20: note: candidate function
    basic_ostream& operator<<(double __f);
                   ^
/usr/local/Cellar/llvm/4.0.1/bin/../include/c++/v1/ostream:219:20: note: candidate function
    basic_ostream& operator<<(long double __f);
                   ^
/usr/local/Cellar/llvm/4.0.1/bin/../include/c++/v1/ostream:790:1: note: candidate function [with _Traits
      = std::__1::char_traits<char>]
operator<<(basic_ostream<char, _Traits>& __os, char __c)
^
/usr/local/Cellar/llvm/4.0.1/bin/../include/c++/v1/ostream:757:1: note: candidate function [with _CharT =
      char, _Traits = std::__1::char_traits<char>]
operator<<(basic_ostream<_CharT, _Traits>& __os, char __cn)
^
/usr/local/Cellar/llvm/4.0.1/bin/../include/c++/v1/ostream:797:1: note: candidate function [with _Traits
      = std::__1::char_traits<char>]
operator<<(basic_ostream<char, _Traits>& __os, signed char __c)
^
/usr/local/Cellar/llvm/4.0.1/bin/../include/c++/v1/ostream:804:1: note: candidate function [with _Traits
      = std::__1::char_traits<char>]
operator<<(basic_ostream<char, _Traits>& __os, unsigned char __c)
^
/usr/local/include/boost/type_traits/detail/has_binary_operator.hpp:70:13: note: candidate function
no_operator operator BOOST_TT_TRAIT_OP (const any&, const any&);
            ^
1 error generated.
make[3]: *** [src/mlpack/tests/CMakeFiles/mlpack_test.dir/imputation_test.cpp.o] Error 1
make[3]: *** Waiting for unfinished jobs....
make[2]: *** [src/mlpack/tests/CMakeFiles/mlpack_test.dir/all] Error 2
make[1]: *** [src/mlpack/tests/CMakeFiles/mlpack_test.dir/rule] Error 2
make: *** [mlpack_test] Error 2

```

rcurtin commented:
Ah I think maybe you are using an older version of Boost than what I have (or, at least, that compiles ok on my system).  But I see what the issue is.  Try changing the line

```
  BOOST_REQUIRE_EQUAL(dm.Type(0), data::Datatype::categorical);
```

to

```
  BOOST_REQUIRE(dm.Type(0) == data::Datatype::categorical);
```

and I think this will solve the problem.

micyril commented:
Yes, it works now.

micyril committed:
Refactor GridSearch to use DatasetMapper

micyril committed:
Use the new GridSearch interface in the HPT module

micyril committed:
Add separate comments

micyril commented:
I have changed the implementation to one that uses `DatasetMapper`. I was forced to take a `DatasetMapper` object as non-const reference in `GridSearch` since `UnmapString` is not marked as a const method (but it seems like it should). If you change it, I can add the changes too.

rcurtin commented:
Yeah, I remember there were problems with `boost::bimap` when I implemented `DatasetMapper` that forced me to not mark `UnmapString()` as `const`.  But in https://github.com/mlpack/mlpack/pull/1093, I removed the `boost::bimap` dependency so `UnmapString()` can be marked `const` finally, which should fix the issue.  We can wait until mlpack#1093 is merged to change it here.

micyril committed:
Add gradient evaluation to CVFunction

micyril committed:
Add support for GradientDescent in the HPT module

micyril committed:
Refactor template conditions for InitAndOptimize

micyril committed:
Add assertion for argument types of Optimize

The added assertion causes a compilation error if one of aguments passed
to the Optimize method of HyperParameterTuner is neither of an
arithmetic type, nor of a collection type, nor fixed with the Fixed
function.

micyril committed:
Add assertion that input collections aren't empty

rcurtin commented on src/mlpack/core/hpt/hpt.hpp:
```diff
@@ -105,6 +106,26 @@ class HyperParameterTuner
+   * is going to be used when it is greater than the increase calculated with
+   * the rules described in the documentation for RelativeDelta().
+   */
+  double& MinDelta() { return minDelta; }
```
It would be good to add `const` versions of these functions so that they can be used in other methods where the `HyperParameterTuner` is `const`.

micyril replied:
Ok. When I designed the interface I assumed that`HyperParameterTuner` objects would not be used with the `const` qualifier.

rcurtin replied:
Users using a `const HyperParameterTuner` seems unlikely, but it's hard to predict what users will do, so in my opinion it's better to add the `const` versions.

rcurtin commented on src/mlpack/core/hpt/cv_function_impl.hpp:
```diff
@@ -65,6 +70,26 @@ template<typename CVType,
+    increasedParameters(i) += delta;
+    gradient(i) = (Evaluate(increasedParameters) - Evaluate(parameters)) /
+        delta;
+    increasedParameters(i) = parameters(i);
```
I think you could cache `Evaluate(parameters)` for a big speed gain.  If you did that, you could also remove `increasedParameters` and just perturb `parameters(i)` directly.

micyril replied:
Oh, sure. I'm surprised why I haven't cached `Evaluate(parameters)` it in the first place. But I guess we should keep `increasedParameters` as it is since we accept `parameters` with the `const` qualifier.

rcurtin replied:
Ah, right, it is necessary to keep `increasedParameters`.

rcurtin reviewed (commented):
Looks great.  If you want to open a PR in the main mlpack repository we can continue the discussions there, but I don't think I have any more comments past the two simple ones I made here.

micyril commented:
I thought that it would be reasonable to open a PR in the main mlpack repository after the patch for DatasetMapper got integrated into the main repository as well - otherwise HyperParameterModule will not work. Do you want me to open a PR anyway?

rcurtin commented:
Do you mean https://github.com/mlpack/mlpack/pull/1093?  If so I will make sure that gets merged shortly.

micyril commented:
No, actually I meant the patch that you provided me earlier (https://gist.github.com/rcurtin/0628ebb53f349ce60a6dc6099902005d). But #1093 is also useful - it will allow to pass `datasetInfo` with the `const` qualifier to `GridSearch` and `GradientDecent`.

rcurtin commented:
Ah, ok.  Couldn't you just include the patch in the PR that you open?

micyril committed:
Add const getters in HyperParameterTuner

micyril committed:
Make CVFunction cache computations for gradient
